src/bson.rs

```diff
@@ -486,184 +595,243 @@ impl Bson {
-                if time >= 0 && increment >= 0 {
-                    return Bson::TimeStamp(TimeStamp {
-                        time: time as u32,
-                        increment: increment as u32,
-                    });
-                }
-            } else if let (Ok(time), Ok(increment)) = (values.get_i64("t"), values.get_i64("i")) {
-                if time >= 0
-                    && time <= std::u32::MAX as i64
-                    && increment >= 0
-                    && increment <= std::u32::MAX as i64
-                {
-                    return Bson::TimeStamp(TimeStamp {
-                        time: time as u32,
-                        increment: increment as u32,
-                    });
-                }
-            } else if let (Ok(hex), Ok(t)) = (values.get_str("$binary"), values.get_i64("type")) {
-                let ttype = t as u8;
-                return Bson::Binary(Binary {
-                    subtype: From::from(ttype),
-                    bytes: hex::decode(hex.as_bytes())
-                        .expect("$binary value is not a valid Hex encoded bytes"),
-                });
-            }
-        } else if values.len() == 1 {
-            if let Ok(code) = values.get_str("$code") {
-                return Bson::JavaScriptCode(code.to_owned());
-            } else if let Ok(hex) = values.get_str("$oid") {
-                return Bson::ObjectId(oid::ObjectId::with_string(hex).unwrap());
-            } else if let Ok(long) = values
-                .get_document("$date")
-                .and_then(|inner| inner.get_i64("$numberLong"))
-            {
-                return Bson::UtcDatetime(
-                    Utc.timestamp(long / 1000, ((long % 1000) * 1_000_000) as u32),
-                );
-            } else if let Ok(sym) = values.get_str("$symbol") {
-                return Bson::Symbol(sym.to_owned());
-            } else if let Ok(undefined) = values.get_bool("$undefined") {
-                if undefined {
-                    return Bson::Undefined;
+            ["$numberDouble"] => match doc.get_str("$numberDouble") {
+                Ok("Infinity") => return Bson::FloatingPoint(f64::INFINITY),
+                Ok("-Infinity") => return Bson::FloatingPoint(f64::NEG_INFINITY),
+                Ok("NaN") => return Bson::FloatingPoint(f64::NAN),
+                Ok(other) => {
+                    if let Ok(d) = other.parse() {
+                        return Bson::FloatingPoint(d);
+                    }
+                }
+                _ => {}
+            },
+
+            #[cfg(feature = "decimal128")]
+            ["$numberDecimal"] => {
+                if let Ok(d) = doc.get_str("$numberDecimal") {
+                    if let Ok(d) = d.parse() {
+                        return Bson::Decimal128(d);
+                    }
+                }
+            }
+
+            ["$binary"] => {
+                if let Ok(binary) = doc.get_document("$binary") {
```

Comment:
Things like this are a bit messy, but I'm hesitant to use serde as part of the extended JSON parsing

Reply:
This is actually trickier than it looks for a few reasons. First, we're trying to get multiple keys from the document, which means we won't be able to use a single long chain. It would look more like this:

```rust
doc.get_document("$binary")
    .and_then(|binary| {
	    let bytes = binary.get_str("base64").and_then(|bytes|base64::decode(bytes));
		let subtype = binary.get_str("subType").and_then(|subtype| hex::decode(subtype));

        // etc.
	})
```

This isn't too bad, except the error types aren't the same; the first error in each subchain is a `bson::ordered::ValueAccessError`, but then they switch to errors from separate crates. `and_then` requires that the same error type is used for for `self` and for the return value. 

We also run into issues with the innermost `if` statement, which isn't checking for `Ok` but for the length being 1. We'd have to construct some sort of error for cases where the subtype is longer than one byte. Finally, we'd still need an `if let` at the top level, since we're not actually returning any error from this branch.

Overall, I'm not sure this is much cleaner than what we have here. I think dealing with all three of the above issues would end up with something that's harder to understand at a glance than what we have now, despite its ugliness. That said, there's still probably room for improvement. My first instinct is to abstract all this out to method that returns `Option<Bson>`, which would let us use `.ok()?` for all the result cases and return `None` if the subtype is not 1. That would let us get away with a single `if let` for the branch inside this method.

Reply:
>My first instinct is to abstract all this out to method that returns Option<Bson>, which would let us use .ok()? for all the result cases and return None if the subtype is not 1.

Yeah that seems like a good idea.

Reply:
lgtm